### PR TITLE
fix: localnet displays a warning when image is out of date

### PR DESCRIPTION
**Pull request by inaie-makerx**
https://dev.azure.com/MakerXStudio/AlgoKit%20(Algorand%20Foundation)/_workitems/edit/431

### CODE REVIEW

File: src/algokit/core/sandbox.py
Comment (daniel-makerx) [inline src/algokit/core/sandbox.py:156]: Rather than loading the docker-compose.yaml contents, could instead make the algod and indexer image tags constants and reference them here and in the `get_docker_compose_yml`

File: src/algokit/core/sandbox.py
Comment (daniel-makerx) [inline src/algokit/core/sandbox.py:165]: Include in this message what the user has to do to get the latest version (run with --update)

File: src/algokit/core/sandbox.py
Comment (daniel-makerx) [inline src/algokit/core/sandbox.py:150]: I think this should return None here if we can't fetch the latest, and then skip the warning if None is returned so that users don't get warnings if we can't determine the latest version

File: tests/localnet/conftest.py
Comment (daniel-makerx) [inline tests/localnet/conftest.py:51]: `localnet_out_of_date` ?

File: tests/localnet/conftest.py
Comment (daniel-makerx) [inline tests/localnet/conftest.py:26]: If we make the image tags constants as mentioned earlier, then they can also be referenced here in the tests instead of defining them in multiple places

File: tests/localnet/conftest.py
Comment (aorumbayev) [inline tests/localnet/conftest.py:21]: Minor detail, i think
`_localnet_up_to_date_` -> `localnet_up_to_date` might be better? 

Its used in at least 2 separate test sets so i assume it makes more sense to remove the `_`. If you look at the only similar example of 'private' fixture in `test_doctor.py` that fixture is only used within a scope of a single file hence makes sense to mark it private there.

File: tests/localnet/conftest.py
Comment (aorumbayev) [inline tests/localnet/conftest.py:48]: In contrast with `localnet_up_to_date`, `_localnet_out_of_date` is only used by a single test case in `test_localnet_start.py`. Is there any need to have it defined in conftest in this case?

File: src/algokit/core/sandbox.py
Comment (aorumbayev) [inline src/algokit/core/sandbox.py:166]:
```suggestion
                "algod has a new version available, run `algokit localnet reset --update` to get the latest version"
```

File: src/algokit/core/sandbox.py
Comment (aorumbayev) [inline src/algokit/core/sandbox.py:160]:
```suggestion
                "indexer has a new version available, run `algokit localnet reset --update` to get the latest version"
```

File: src/algokit/core/sandbox.py
Comment (aorumbayev) [inline src/algokit/core/sandbox.py:134]: output has a newline character at the end which isn't accounted for, can result in a bug where it forgets to strip the rightmost `]` symbol. See if there are better ways to get just the sha part, maybe --format flag has some further ways to parse the output on docker side instead? Otherwise make sure to run strip() on the output before further operations

Comment (aorumbayev) [reply src/algokit/core/sandbox.py:134]: Please also make sure to also reflect the real outputs in the mocks, current tests aren't catching scenarios like that because mocked return values are already trimmed. Potentially adding extra asserts where applicable to ensure the sha strings were correctly parsed and etc (there are no such test checks at the moment)

File: tests/localnet/test_localnet_start.py
Comment (aorumbayev) [inline tests/localnet/test_localnet_start.py:183]: Might be useful to add few more tests testing edge cases when you have local docker inspect invocation failing on some exception (should be handled without raising any unhandled exceptions). Similarly testing a scenario where remote local inspect throws an error, as expected though it should print a debug warning but user of the cli shouldn't see any errors/logs and instead its just continues with the execution of the command.

**Review by aorumbayev — CHANGES_REQUESTED** [review]

File: tests/localnet/conftest.py
Comment (inaie-makerx) [reply tests/localnet/conftest.py:21]: I added the underscore because of this error 
Fixture `localnet_up_to_date` does not return anything, add leading underscoreRuff(PT004)

File: src/algokit/core/sandbox.py
Comment (inaie-makerx) [reply src/algokit/core/sandbox.py:134]: I have changed to use the .Id it seems to return the sha formatted on the way we need already

File: tests/localnet/conftest.py
Comment (aorumbayev) [reply tests/localnet/conftest.py:21]: I see, looks like ruff is enforcing pytest conventions, ignore this one then all good

File: src/algokit/core/sandbox.py
Comment (aorumbayev) [reply src/algokit/core/sandbox.py]: @inaie-makerx I'm sorry you can't really use ID in this case case RepoDigest vs ID are derived in different manners and we rely on repo digest on remote image side. All i meant was that when you run algokit cli (aside from mocks in the tests) this inspect response returns a newline character (happens in current state with ID as well). 

Can you please:
1. Make sure the output is stripper/trimmed from any special characters on left/right side before you return it from `_get_local_image_version` ?
2. Change the ID back to RepoDigest? (see [ref](https://stackoverflow.com/questions/56364643/whats-the-difference-between-a-docker-images-image-id-and-its-digest))

Here is what you can do as an option, docker relies on go under the hood so you can rely on go templating capabilities to do something like this:

```python
    def _get_local_image_version(self, image_name: str) -> str | None:
        """
        Get the local version of a Docker image
        """
        try:
            arg = '{{index (split (index .RepoDigests 0) "@") 1}}'
            local_version = run(
                ["docker", "image", "inspect", image_name, "--format", arg],
                cwd=self.directory,
                bad_return_code_error_message="Failed to get image inspect",
            )
            # Remove brackets and split on '@' symbol and get the SHA hash
            return local_version.output.strip()
        except Exception:
            return None
```

**Review by aorumbayev — CHANGES_REQUESTED** [review]
@inaie-makerx unresolved one of the comments as the bug still persists, please address and I'll approve if all good.

File: tests/localnet/test_localnet_start.py
Comment (aorumbayev) [inline tests/localnet/test_localnet_start.py:26]: Should be just `sha256:bbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbb\n` without square brackets due to `{{index (split (index .RepoDigests 0) "@") 1}}`

File: tests/localnet/test_localnet_start.py
Comment (aorumbayev) [inline tests/localnet/test_localnet_start.py:31]: Same here,  should be just `sha256:aa....\n`

**Review by aorumbayev — APPROVED** [review]
@inaie-makerx approving but please address the last comment with regards to values you are mocking for inspect command. Before merging, please make sure that conventional commit prefix is set to `fix` instead of a `feat` as this is more of a minor enhancement into existing localnet features so a patch bump is more appropriate than a minor bump. 

@daniel-makerx any chance you can take one extra look as well so we get second approval for this PR?

**Review by robdmoore — APPROVED** [review]

**Comment by robdmoore** [comment]
Nice!

File: tests/localnet/test_localnet_start.py
Comment (aorumbayev) [reply tests/localnet/test_localnet_start.py]: These are still in place? Am i missing something? I think you pushed a commit to remove them from the test but unsure if these are still correct, the output should not contain [{title}/{image title}@sha256:{hash}] = we parse it out via go templating in format param.

File: tests/localnet/test_localnet_start.py
Comment (aorumbayev) [reply tests/localnet/test_localnet_start.py]: These are still in place? Am i missing something? I think you pushed a commit to remove them from the test but unsure if these are still correct, the output should not contain [{title}/{image title}@sha256:{hash}] = we parse it out via go templating in format param.

**Pull request merged**
